pilot/pkg/serviceregistry/kube/controller/ambient/sidecar_interop.go

```diff
@@ -0,0 +1,137 @@
+	ServicesByAddress krt.Index[networkAddress, model.ServiceInfo],
+) {
+	// When sending information to a workload, we need two bits of information:
+	// * Does it support tunnel? If so, we will need to use HBONE
+	// * Does it have a waypoint? if so, we will need to send to the waypoint
+	// Record both of these.
+	ServiceEds := krt.NewCollection(
+		Services,
+		func(ctx krt.HandlerContext, svc model.ServiceInfo) *serviceEDS {
+			wp := svc.Service.Waypoint
+			if wp == nil {
+				return nil
+			}
+			var waypointServiceKey string
+			switch addr := wp.Destination.(type) {
+			// Easy case: waypoint is already a hostname. Just return it directly
+			case *workloadapi.GatewayAddress_Hostname:
+				hn := addr.Hostname
+				waypointServiceKey = hn.Namespace + "/" + hn.Hostname
+			// Hard case: waypoint is an IP address. Need to look it up.
+			case *workloadapi.GatewayAddress_Address:
+				wAddress := addr.Address
+				serviceKey := networkAddress{
+					network: wAddress.Network,
+					ip:      mustByteIPToString(wAddress.Address),
+				}
+				waypointSvc := krt.FetchOne(ctx, Services, krt.FilterIndex(ServicesByAddress, serviceKey))
+				if waypointSvc == nil {
+					return nil
+				}
+				waypointServiceKey = waypointSvc.ResourceName()
+			}
+			workloads := krt.Fetch(ctx, Workloads, krt.FilterIndex(WorkloadsByServiceKey, waypointServiceKey))
+			return &serviceEDS{
+				ServiceKey: svc.ResourceName(),
+				WaypointInstance: slices.Map(workloads, func(e model.WorkloadInfo) *workloadapi.Workload {
+					return e.Workload
+				}),
+			}
+		},
+		krt.WithName("ServiceEds"))
+	ServiceEds.RegisterBatch(
+		PushXds(xdsUpdater, func(svc serviceEDS) model.ConfigKey {
+			ns, hostname, _ := strings.Cut(svc.ServiceKey, "/")
+			return model.ConfigKey{Kind: kind.ServiceEntry, Name: hostname, Namespace: ns}
+		}), false)
+}
+
+func (a *index) ServicesWithWaypoint(key string) []model.ServiceWaypointInfo {
+	res := []model.ServiceWaypointInfo{}
+	var svcs []model.ServiceInfo
+	if key == "" {
+		svcs = a.services.List()
+	} else {
+		svcs = ptr.ToList(a.services.GetKey(krt.Key[model.ServiceInfo](key)))
+	}
+	for _, s := range svcs {
+		wp := a.waypoints.GetKey(krt.Key[Waypoint](s.Waypoint.ResourceName))
+		if wp == nil {
+			continue
+		}
+		wi := model.ServiceWaypointInfo{
+			Service: s.Service,
+		}
+		switch addr := wp.GetAddress().GetDestination().(type) {
+		// Easy case: waypoint is already a hostname. Just return it directly
+		case *workloadapi.GatewayAddress_Hostname:
+			wi.WaypointHostname = addr.Hostname.Hostname
+		// Hard case: waypoint is an IP address. Need to look it up.
+		case *workloadapi.GatewayAddress_Address:
+			wpAddr, _ := netip.AddrFromSlice(addr.Address.Address)
+			wpNetAddr := networkAddress{
+				network: addr.Address.Network,
+				ip:      wpAddr.String(),
+			}
+			waypoints := a.services.ByAddress.Lookup(wpNetAddr)
+			if len(waypoints) == 0 {
+				// No waypoint found.
+				continue
+			}
```

Comment:
why don't we set `wi.WaypointHostname` to the service's hostname here?

Reply:
What we want to build is something like

```
model.ServiceWaypointInfo {
  Service <httpbin service>
  WaypointHostname "my-waypoint.svc.cluster.local
}
```

So the hostname is the waypoints, not the services

Reply:
This is just broken -- lack of tests due to this being a pretty obscure code path. I added tests which caught this + fixed it